src/dmd/expressionsem.d

```diff
@@ -10897,6 +10897,39 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
         if (exp.e1.op == TOK.type || exp.e2.op == TOK.type)
         {
+            /* https://issues.dlang.org/show_bug.cgi?id=12520
+             * empty tuples are represented as types so special cases are added
+             * so that they can be compared for equality with tuples of values.
+             */
+
+            // cast e1 and e2 when they are empty tuples or tuples made only of types
+            TypeTuple tt1 = exp.e1.op == TOK.type ? exp.e1.isTypeExp().type.isTypeTuple() : null;
+            TypeTuple tt2 = exp.e2.op == TOK.type ? exp.e2.isTypeExp().type.isTypeTuple() : null;
+            // only empty type tuples are considered comparable with TupleExps
+            const bool tt1Empty = tt1 && ((tt1.arguments && !tt1.arguments.dim) || !tt1.arguments);
+            const bool tt2Empty = tt2 && ((tt2.arguments && !tt2.arguments.dim) || !tt2.arguments);
+            // cast e1 and e2 when they are tuples made of at least a value
+            TupleExp te1 = (tt1 is null && exp.e1.op == TOK.tuple) ? exp.e1.isTupleExp() : null;
```

Comment:
This is a bit unnecessary. `isTupleExp` returns `null` if `exp.e1` is not a tuple exp, so essentially you're doing the check twice for no benefit.